radix-engine-common/src/data/scrypto/custom_validation.rs

```diff
@@ -1,15 +1,185 @@
 use super::*;
+use crate::types::{NodeId, PackageAddress};
 use crate::*;
 use sbor::rust::prelude::*;
 use sbor::traversal::TerminalValueRef;
 use sbor::*;
 
-impl ValidatableCustomTypeExtension<()> for ScryptoCustomTypeExtension {
-    fn apply_custom_type_validation<'de>(
-        _: &Self::CustomTypeValidation,
+pub trait TypeInfoLookup {
+    fn get_node_type_info(&self, node_id: &NodeId) -> Option<TypeInfoForValidation>;
+    // radix-engine-common is not privy to specifics of engine blueprints
+    // - so instead, we delegate this to the TypeInfoLookup implementation
+    fn is_bucket(&self, type_info: &TypeInfoForValidation) -> bool;
+    fn is_proof(&self, type_info: &TypeInfoForValidation) -> bool;
+}
+
+#[derive(Debug, Clone)]
+pub enum TypeInfoForValidation {
```

Comment:
I started with this approach when implementing stateful custom type validation, but realised it's better, from a dependency perspective, if we move stateful validation into system, so `radix-engine-common` has no dependency on type info enum, address/blueprint constants and node property definition.

Reply:
Let's stick with what's implemented in this PR now. Don't think splitting the validation gives us much benefit now.

Reply:
Ahh sorry I didn't see this comment in time, and it was easy to do so I've separated it back out, PR ready for re-review.

Reply:
No problem